engine/src/main/java/org/terasology/world/internal/EntityAwareWorldProvider.java

```diff
@@ -189,76 +181,55 @@ private void updateBlockEntity(EntityRef blockEntity, Vector3i pos, Block oldTyp
             updateBlockEntityComponents(blockEntity, oldType, type, retainComponents);
         }
 
-        OnChangedBlock changedEvent = new OnChangedBlock(JomlUtil.from(pos), type, oldType);
-        EntityRef regionEntity = blockRegionLookup.get(JomlUtil.from(pos));
+        OnChangedBlock changedEvent = new OnChangedBlock(pos, type, oldType);
+        EntityRef regionEntity = blockRegionLookup.get(pos);
         if (regionEntity != null) {
             regionEntity.send(changedEvent);
         }
         blockEntity.send(changedEvent);
     }
 
     @Override
-    public EntityRef setPermanentBlockEntity(Vector3i blockPosition, EntityRef blockEntity) {
+    public EntityRef setPermanentBlockEntity(Vector3ic blockPosition, EntityRef blockEntity) {
         if (GameThread.isCurrentThread()) {
             EntityRef oldEntity = getExistingBlockEntityAt(blockPosition);
-            blockEntityLookup.put(blockPosition, blockEntity);
+            blockEntityLookup.put(new Vector3i(blockPosition), blockEntity);
```

Comment:
I'm having a hard time working out where it's appropriate to add allocations. Why do we copy blockPosition here? Is it because the Vector3ic we get isn't really immutable, that's only a promise that _this method_ doesn't mutate it?

Reply:
yea, Vector3ic is a promise that the variable is not mutated. so you have to create a new Vector3i so another party doesn't modify the vector.